### PR TITLE
Allow build to standalone branches

### PR DESCRIPTION
**Pull request by shvlv**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added/updated (for bug fixes/features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature.

**What is the current behavior?** (You can also link to an open issue here)
The "Build and push assets" workflow pushes compiled assets back to the development branch. It causes merging conflicts and makes development harder. Tags always belong to detached commits so it is not possible to see linear history with tags.

**What is the new behavior (if this is a feature change)?**
Two new inputs are provided `BUILT_BRANCH_SUFFIX` and `RELEASE_BRANCH_NAME` to allow building and push compiled assets to separated branches. 

Without BUILT_BRANCH_SUFFIX and RELEASE_BRANCH_NAME the current Build and push assets functionality is still in place.

With BUILT_BRANCH_SUFFIX on a branch event compiled assets go to a suffixed branch (like `main-built`).

With RELEASE_BRANCH_NAME on a tag event compiled assets go to the release branch (if the tag points to the last commit of the default GitHub repository branch).

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No.

**Other information**:

### CODE REVIEW

File: docs/build-and-push-assets.md [hunk 1]
```diff
@@ -211,3 +239,14 @@ workflow can install these packages.
...
+---
+
+> What version should I use for Composer?
+
+For tags the pushed tags name are always used.
+
+For branches, it depends on `BUILT_BRANCH_SUFFIX` input value. If the input is not provided
+you can use branch name (i.e. `dev-main` for `main` branch) as usually. If suffix was defined
+the built branch should be used. I.e. for `BUILT_BRANCH_SUFFIX` equals `-built` and branch `main`
+the `dev-main-built` branch should be used as the package version.
```
Comment (tyrann0us) [inline docs/build-and-push-assets.md:252]:
```suggestion
> What version should I use when requiring the package with Composer?

For tags, the pushed tag name is always used.

For branches, it depends on the `BUILT_BRANCH_SUFFIX` input value. If the input is not provided,
you can use the branch name (i.e., `dev-main` for the `main` branch) as usual. If a suffix was defined,
the built branch should be used. I.e., when `BUILT_BRANCH_SUFFIX` is `-built` and branch `main`,
the `dev-main-built` branch should be used as the package version.
```

**Review by tyrann0us — COMMENTED** [review]
Thank you for working on this! The workflow is complex, and I can't assess if there's any room for improvement. Left some comments regarding the documentation, though.

File: .github/workflows/build-and-push-assets.yml [hunk 2]
```diff
@@ -61,7 +81,49 @@ on:
         required: false
 
 jobs:
+  checks:
+    runs-on: ubuntu-latest
+    env:
+      GH_TOKEN: ${{ github.token }}
```
Comment (Chrico) [inline .github/workflows/build-and-push-assets.yml:87]: For what is this `env.GH_TOKEN` used? Please leave a comment after the assignment for "later". :)

File: docs/build-and-push-assets.md [hunk 3]
```diff
@@ -6,15 +6,33 @@ To achieve that, the reusable workflow:
...
 1. installs dependencies defined in `package.json`
 2. builds the package's assets via a build script (see below)
-3. pushes built assets back to the repository
+3. pushes built assets back to the repository (to the same branch or a defined branch)
+
+## Where are assets stored
+
+Two inputs can be used to define branches as assets storage - `BUILT_BRANCH_SUFFIX` and `RELEASE_BRANCH_NAME`.
+
+`BUILT_BRANCH_SUFFIX` is used only for push to branch events. If defined, compiled assets
```
Comment (tyrann0us) [inline docs/build-and-push-assets.md:15]:
```suggestion
`BUILT_BRANCH_SUFFIX` is used only for push-to-branch events. If defined, compiled assets
```

File: .github/workflows/build-and-push-assets.yml [hunk 4]
```diff
@@ -48,6 +48,26 @@ on:
         default: './assets'
         required: false
         type: string
+      BUILT_BRANCH_SUFFIX:
+        description: Suffix for branch creating for compiled assets (`-built` is recommended for convention).
```
Comment (tyrann0us) [inline .github/workflows/build-and-push-assets.yml:52]: Can we use the description from the documentation?
```suggestion
        description: Suffix to calculate the target branch for pushing assets on the `branch` event.
```

File: docs/build-and-push-assets.md [hunk 5]
```diff
@@ -73,17 +96,22 @@ This is not the simplest possible example, but it showcases all the recommendati
...
+| `WORKING_DIRECTORY`   | `'./'`                        | Working directory path                                                                                                          |
+| `COMPILE_SCRIPT_PROD` | `'build'`                     | Script added to `npm run` or `yarn` to build production assets                                                                  |
+| `COMPILE_SCRIPT_DEV`  | `'build:dev'`                 | Script added to `npm run` or `yarn` to build development assets                                                                 |
+| `MODE`                | `''`                          | Mode for compiling assets (`prod` or `dev`)                                                                                     |
+| `ASSETS_TARGET_PATHS` | `'./assets'`                  | Target path(s) for compiled assets                                                                                              |
+| `BUILT_BRANCH_SUFFIX` | `''`                          | Suffix to calculate target branch for pushing assets on `branch` event                                                          |
```
Comment (tyrann0us) [inline docs/build-and-push-assets.md:110]: Please fix the Markdown table syntax after incorporating the changes.
```suggestion
| `BUILT_BRANCH_SUFFIX` | `''`                          | Suffix to calculate the target branch for pushing assets on the `branch` event                                                          |
```

File: docs/build-and-push-assets.md [hunk 6]
```diff
@@ -73,17 +96,22 @@ This is not the simplest possible example, but it showcases all the recommendati
...
+| `RELEASE_BRANCH_NAME` | `''`                          | Target branch for pushing compiled assets on `tag` event                                                                        |
```
Comment (tyrann0us) [inline docs/build-and-push-assets.md:111]:
```suggestion
| `RELEASE_BRANCH_NAME` | `''`                          | Target branch for pushing compiled assets on the `tag` event                                                                        |
```

**Review by tyrann0us — COMMENTED** [review]
Left a couple of minor, mostly wording-related suggestions.

File: .github/workflows/build-and-push-assets.yml [hunk 7]
```diff
@@ -168,26 +250,30 @@ jobs:
...
 
-      - name: Git push
-        if: ${{ env.NO_CHANGES != 'yes' }}
+      - name: Git push for branch
+        if: ${{ github.ref_type == 'branch' && (env.NO_CHANGES != 'yes' || inputs.BUILT_BRANCH_SUFFIX != '') }}
+        run: git push
+
+      - name: Git push for tag
+        if: ${{ github.ref_type == 'tag' && (env.NO_CHANGES != 'yes' || env.RELEASE_BRANCH_ENABLED == 'yes') }}
         run: git push
```
[the remark this answers is not in the document]
Comment (shvlv) [reply .github/workflows/build-and-push-assets.yml:266]: yes, we have to make our "built" branch and "release" branches in sync with the main branch. So we have to merge and push even if there are no new commits (compiled assets).

File: .github/workflows/build-and-push-assets.yml [hunk 8]
```diff
@@ -48,6 +48,26 @@ on:
         default: './assets'
         required: false
         type: string
+      BUILT_BRANCH_SUFFIX:
+        description: Suffix for branch creating for compiled assets (`-built` is recommended for convention).
+        type: string
+        default: ''
+        required: false
+      RELEASE_BRANCH_NAME:
+        description: Branch name where consequentially tags are moved.
```
[the remark this answers is not in the document]
Comment (tyrann0us) [reply .github/workflows/build-and-push-assets.yml:57]: What about this?
```suggestion
        description: On tag events, target branch where compiled assets are pushed and the tag is moved to.
```
If you agree, please also update the documentation. Thanks!

**Review by tyrann0us — APPROVED** [review]
LGTM. Thanks for working on this!

**Pull request merged**